File: ferric_continuum/foundation/smart_pointers_demo.cc
#include "absl/log/globals.h"
#include "absl/log/initialize.h"
#include "absl/log/log.h"

#include "smart_pointers.hh"

namespace ff = ferric::foundation;

int main() {
  // Initialize Abseil logging
  absl::InitializeLog();
  absl::SetStderrThreshold(absl::LogSeverityAtLeast::kInfo);

  LOG(INFO) << "=== Smart Pointers & RAII Demo ===";

  // ===== unique_ptr - Exclusive Ownership =====
  LOG(INFO) << "1. unique_ptr - Exclusive Ownership";
  {
    auto list = ff::create_list({1, 2, 3, 4, 5});
    LOG(INFO) << "   Created list with " << ff::count_nodes(list.get()) << " nodes";
    LOG(INFO) << "   First value: " << list->value();

    // Transfer ownership
    auto list2 = std::move(list);
    LOG(INFO) << "   After move:";
    LOG(INFO) << "   - list is nullptr: " << (list == nullptr);
    LOG(INFO) << "   - list2 owns the data: " << (list2 != nullptr);

    LOG(INFO) << "   Leaving scope - automatic cleanup!";
  }  // list2 destroyed here, all nodes automatically deleted
  LOG(INFO) << "   All memory cleaned up automatically.";

  // ===== shared_ptr - Shared Ownership =====
  LOG(INFO) << "2. shared_ptr - Shared Ownership";
  {
    LOG(INFO) << "   Resources alive: " << ff::Resource::instance_count();

    auto resource = ff::create_shared_resource(42);
    LOG(INFO) << "   Created resource " << resource->id();
    LOG(INFO) << "   Use count: " << resource.use_count();
    LOG(INFO) << "   Resources alive: " << ff::Resource::instance_count();

    {
      // Share ownership
      auto shared = ff::share_resource(resource, 3);
      LOG(INFO) << "   After sharing with 3 more owners:";
      LOG(INFO) << "   Use count: " << resource.use_count();
      LOG(INFO) << "   Resources alive: " << ff::Resource::instance_count();

      LOG(INFO) << "   Leaving inner scope...";
    }  // shared vector destroyed, but resource still alive

    LOG(INFO) << "   Back to outer scope:";
    LOG(INFO) << "   Use count: " << resource.use_count();
    LOG(INFO) << "   Resources alive: " << ff::Resource::instance_count();

    LOG(INFO) << "   Leaving outer scope...";
  }  // resource destroyed when last shared_ptr goes away
  LOG(INFO) << "   Resources alive: " << ff::Resource::instance_count();

  // ===== RAII Pattern =====
  LOG(INFO) << "3. RAII Pattern";
  {
    ff::FileGuard file("data.txt");
    LOG(INFO) << "   File opened: " << file.filename();
    LOG(INFO) << "   Is open: " << file.is_open();

    // Transfer ownership
    ff::FileGuard file2 = std::move(file);
    LOG(INFO) << "   After move:";
    LOG(INFO) << "   - file is closed: " << (!file.is_open());
    LOG(INFO) << "   - file2 owns resource: " << file2.is_open();

    LOG(INFO) << "   Leaving scope...";
  }  // file2 destroyed, file automatically closed
  LOG(INFO) << "   File automatically closed in destructor.";

  LOG(INFO) << "Key Benefits:";
  LOG(INFO) << "- No manual delete needed - automatic cleanup";
  LOG(INFO) << "- Exception-safe - cleanup happens even if exception thrown";
  LOG(INFO) << "- Clear ownership semantics - who owns what?";
  LOG(INFO) << "- unique_ptr: exclusive ownership (can't be copied)";
  LOG(INFO) << "- shared_ptr: shared ownership (reference counted)";
  LOG(INFO) << "- RAII: resource lifetime tied to object lifetime";

  return 0;
}

File: ferric_continuum/foundation/smart_pointers_demo.rs
use smart_pointers::{
    count_nodes, create_list, share_resource, Counter, FileGuard, Resource,
};
use std::rc::Rc;
use tracing::{info, Level};
use tracing_subscriber;

fn main() {
    // Initialize logging
    tracing_subscriber::fmt().with_max_level(Level::INFO).init();

    info!("=== Smart Pointers & RAII Demo (Rust) ===");

    // ===== Box - Exclusive Ownership =====
    info!("1. Box<T> - Exclusive Ownership");
    {
        let list = create_list(&[1, 2, 3, 4, 5]);
        let node_count = count_nodes(list.as_ref().map(|b| b.as_ref()));
        info!(node_count, "Created list");

        if let Some(ref node) = list {
            info!(value = node.value(), "First value");
        }

        // Transfer ownership
        let list2 = list; // list moved
        info!("After move:");
        info!("- list is moved (no longer accessible)");
        info!(owns_data = list2.is_some(), "list2 owns the data");

        info!("Leaving scope - automatic cleanup!");
    } // list2 dropped, all nodes automatically freed
    info!("All memory cleaned up automatically.");

    // ===== Rc - Shared Ownership =====
    info!("2. Rc<T> - Shared Ownership");
    {
        Resource::reset_count();
        info!(alive = Resource::instance_count(), "Resources alive");

        let resource = Resource::new(42);
        info!(id = resource.id(), "Created resource");
        info!(count = Rc::strong_count(&resource), "Strong count");
        info!(alive = Resource::instance_count(), "Resources alive");

        {
            // Share ownership
            let _shared = share_resource(Rc::clone(&resource), 3);
            info!("After sharing with 3 more owners:");
            info!(count = Rc::strong_count(&resource), "Strong count");
            info!(alive = Resource::instance_count(), "Resources alive");

            info!("Leaving inner scope...");
        } // _shared vector dropped

        info!("Back to outer scope:");
        info!(count = Rc::strong_count(&resource), "Strong count");
        info!(alive = Resource::instance_count(), "Resources alive");

        info!("Leaving outer scope...");
    } // resource dropped when last Rc goes away
    info!(alive = Resource::instance_count(), "Resources alive");

    // ===== RAII Pattern =====
    info!("3. RAII Pattern (Drop trait)");
    {
        let file = FileGuard::new("data.txt");
        info!(filename = file.filename(), "File opened");
        info!(is_open = file.is_open(), "File status");

        // Transfer ownership with move
        let file2 = file; // file moved to file2
        info!("After move:");
        info!(owns_resource = file2.is_open(), "file2 owns resource");

        info!("Leaving scope...");
    } // file2 dropped, Drop::drop called automatically
    info!("File automatically closed in Drop::drop.");

    // ===== Interior Mutability =====
    info!("4. Interior Mutability (RefCell)");
    {
        let counter = Counter::new();
        info!(value = counter.get(), "Initial value");

        // Multiple references can mutate through RefCell
        let c1 = Rc::clone(&counter);
        let c2 = Rc::clone(&counter);

        c1.increment();
        c2.increment();
        counter.increment();

        info!(value = counter.get(), "After 3 increments");
        info!(count = Rc::strong_count(&counter), "Strong count");
    }

    info!("Key Differences:");
    info!("- C++: Manual smart pointer selection (unique_ptr vs shared_ptr)");
    info!("- Rust: Ownership enforced at compile time");
    info!("- Box<T>: Like unique_ptr (exclusive ownership)");
    info!("- Rc<T>: Like shared_ptr (reference counted)");
    info!("- Arc<T>: Thread-safe Rc (atomic reference counting)");
    info!("- Drop trait: Automatic cleanup (like RAII/destructors)");
    info!("- RefCell: Interior mutability when needed");

    info!("Rust's Advantages:");
    info!("- Compiler prevents use-after-free at compile time");
    info!("- No null pointer dereferencing in safe code");
    info!("- Memory leaks harder to create (but still possible with Rc cycles)");
}
